lib/src/fake_aggregate_query.dart

```diff
@@ -1,19 +1,128 @@
 import 'package:cloud_firestore/cloud_firestore.dart';
+import 'package:cloud_firestore_platform_interface/cloud_firestore_platform_interface.dart'
+    hide AggregateQuery;
+import 'package:cloud_firestore_platform_interface/cloud_firestore_platform_interface.dart'
+    as platform_interface;
+import 'package:collection/collection.dart';
+import 'package:flutter/foundation.dart';
 
 import 'fake_aggregate_query_snapshot.dart';
 
 class FakeAggregateQuery implements AggregateQuery {
   final Query _query;
+  final Iterable<AggregateField?> _aggregateFields;
 
-  FakeAggregateQuery(this._query);
+  FakeAggregateQuery(this._query, [this._aggregateFields = const []]);
 
   @override
   Future<AggregateQuerySnapshot> get(
       {AggregateSource source = AggregateSource.server}) async {
     final snapshot = await _query.get();
-    return FakeAggregateQuerySnapshot(_query, snapshot.size);
+    final delegate = _getAggregateQuerySnapshotPlatform(snapshot: snapshot);
+    return FakeAggregateQuerySnapshot(_query, delegate);
   }
 
   @override
   Query<Object?> get query => _query;
+
+  @override
+  AggregateQuery count() => _query.count();
+
+  AggregateQuerySnapshotPlatform _getAggregateQuerySnapshotPlatform({
+    required QuerySnapshot<Object?> snapshot,
+  }) {
+    final dataMaps = snapshot.docs.map((e) => e.data() as Map<String, dynamic>);
+    final delegate = AggregateQuerySnapshotPlatform(
+      count: snapshot.size,
+      sum: buildAggregateQueryResponseList(
+        dataMaps: dataMaps,
+        aggregateFields: _aggregateFields,
+        aggregateType: AggregateType.sum,
+      ),
+      average: buildAggregateQueryResponseList(
+        dataMaps: dataMaps,
+        aggregateFields: _aggregateFields,
+        aggregateType: AggregateType.average,
+      ),
+    );
+    return delegate;
+  }
+
+  @visibleForTesting
+  static List<AggregateQueryResponse> convertValuesMapToResponseList(
+    Map<String, double> keyNumMap,
+    AggregateType aggregateType,
+  ) {
+    return keyNumMap.entries
+        .map((e) => AggregateQueryResponse(
+              type: aggregateType,
+              value: e.value,
+              field: e.key,
+            ))
+        .toList();
+  }
+
+  @visibleForTesting
+  static Iterable<AggregateField> getFieldsWithType({
+    required Iterable<AggregateField?> fields,
+    required AggregateType type,
+  }) {
+    final nonNullFields = fields.whereNotNull();
+    switch (type) {
```

Comment:
I tried using Dart 3.0's new Switch expressions but sadly it didn't work because the value wasn't a type 😞 .

```dart
final fieldType = switch (type) {
  AggregateType.sum => platform_interface.sum,
  AggregateType.average => platform_interface.average,
  AggregateType.count => platform_interface.count,
  _ => throw UnimplementedError('Unknown AggregateType: $type')
};
final nonNullFields = fields.whereNotNull();
// compiler says: The name 'fieldType' isn't a type, so it can't be used as a type argument.
// Try correcting the name to an existing type, or defining a type named 'fieldType'.
return nonNullFields.whereType<fieldType>();
```

https://dart.dev/language/branches#switch-expressions

Reply:
I did a little search about it, seems we cannot pass the value as Generic Type
https://stackoverflow.com/questions/66021661/how-can-i-cast-a-variable-to-generic-type-in-dart

Reply:
btw, maybe we could use where instead:
```dart
return nonNullFields.where((e) => e.runtimeType == fieldType);
```

Reply:
> I tried using Dart 3.0's new Switch expressions but sadly it didn't work because the value wasn't a type 😞 .
> 
> ```dart
> final fieldType = switch (type) {
>   AggregateType.sum => platform_interface.sum,
>   AggregateType.average => platform_interface.average,
>   AggregateType.count => platform_interface.count,
>   _ => throw UnimplementedError('Unknown AggregateType: $type')
> };
> final nonNullFields = fields.whereNotNull();
> // compiler says: The name 'fieldType' isn't a type, so it can't be used as a type argument.
> // Try correcting the name to an existing type, or defining a type named 'fieldType'.
> return nonNullFields.whereType<fieldType>();
> ```
> 
> https://dart.dev/language/branches#switch-expressions

That's good to simplify the code which makes it easier to read. Nice!

Reply:
> btw, maybe we could use `where` instead:
> 
> ```dart
> return nonNullFields.where((e) => e.runtimeType == fieldType);
> ```

Oh! Good idea